File: braindance/examples/streaming_workshop/catalog_workspace.js
// Catalog metadata and local experiment selection share the Analysis workspace.
(()=>{
  const root=views.catalog;
  root.insertAdjacentHTML('beforeend',`
    <p class="help">Browse recordings and saved experiments, then select one for analysis or playback.</p>
    <div class="catalog-summary">
      <section class="panel"><h2>Current experiment</h2><p id="catalogCurrentStatus" role="status"></p><p id="catalogCurrentPath" class="help"></p><div class="toolbar"><button id="catalogCurrent">Analyze current experiment</button><button id="catalogMonitor">Live recording</button></div></section>
      <section class="panel"><h2>Selected for analysis</h2><p id="catalogSelected">No experiment selected.</p><button id="catalogAnalysis">Open Analysis</button></section>
    </div>
    <section class="panel">
      <div class="toolbar"><h2>Experiments & recordings</h2><button id="catalogRefresh">Refresh catalog</button></div>
      <p class="help">Refresh reloads the configured CSV and discovers saved workshop runs and tutorial experiments. Add a parent folder to discover the experiments inside it; added paths are saved in the workshop index.</p>
      <form id="catalogAddForm" class="toolbar"><label class="analysis-path">Experiment or parent directory, experiment JSON, or recording<input id="catalogAddPath" placeholder="/path/to/experiment" required></label><button id="catalogAdd" type="submit">Add experiment</button></form>
      <div class="toolbar"><label class="analysis-path">Search catalog<input id="catalogSearch" type="search" placeholder="Experiment, project, chip, type or path"></label><label>Source<select id="catalogSource"><option value="">All sources</option><option>Configured catalog</option><option>Workshop run</option><option>Tutorial data</option><option>Added locally</option></select></label></div>
      <p id="catalogMessage" role="status" aria-live="polite"></p><p id="catalogWarnings" class="help"></p>
      <div class="catalog-table"><table><thead><tr><th>Experiment</th><th>Project / chip</th><th>Type / source</th><th>Actions</th></tr></thead><tbody id="catalogRows"></tbody></table></div>
      <div class="toolbar"><button id="catalogPrevious">Previous</button><span id="catalogCount"></span><button id="catalogNext">Next</button></div>
      <p id="catalogLocation" class="help"></p>
    </section>`);
  const el=id=>document.getElementById(id);
  let entries=[], selectedPath='', page=0, busy=false, loaded=false;
  function render(){
    const query=el('catalogSearch').value.toLowerCase(), source=el('catalogSource').value;
    const filtered=entries.filter(entry=>(!source||entry.source===source)&&[entry.name,entry.project,entry.chip,entry.kind,entry.path].join(' ').toLowerCase().includes(query));
    page=Math.min(page,Math.max(0,Math.ceil(filtered.length/100)-1));
    const rows=filtered.slice(page*100,(page+1)*100).map(entry=>{
      const row=document.createElement('tr');
      const name=document.createElement('td'), title=document.createElement('strong'), path=document.createElement('small');
      title.textContent=entry.name;path.textContent=entry.path||'No local path';name.append(title,path);
      const project=document.createElement('td');project.textContent=[entry.project,entry.chip].filter(Boolean).join(' / ')||'—';
      const source=document.createElement('td');source.textContent=[entry.kind,entry.source].filter(Boolean).join(' · ');
      const action=document.createElement('td'), button=document.createElement('button');
      button.textContent=selectedPath===entry.path?'Selected · open Analysis':'Select for analysis';button.disabled=!entry.available||busy;
      button.onclick=()=>select('path',entry.path);action.append(button);
      if(entry.available){
        const replay=document.createElement('button');replay.textContent='Load playback';replay.disabled=busy||running();
        replay.onclick=async()=>{
          if(await window.workshopPlayback.load(entry.path))showView('experiment');
        };action.append(replay);
      }
      if(!entry.available){const reason=document.createElement('small');reason.textContent=entry.reason;action.append(reason);}
      row.append(name,project,source,action);return row;
    });
    if(!rows.length){const row=document.createElement('tr'), cell=document.createElement('td');cell.colSpan=4;cell.textContent=entries.length?'No matching experiments.':'No experiments yet. Run an experiment or add a local path above.';row.append(cell);rows.push(row);}
    el('catalogRows').replaceChildren(...rows);
    el('catalogCount').textContent=`${filtered.length} entries${filtered.length?` · page ${page+1} of ${Math.ceil(filtered.length/100)}`:''}`;
    el('catalogPrevious').disabled=page===0;el('catalogNext').disabled=(page+1)*100>=filtered.length;
  }
  function current(){
    const snapshot=state||{};
    el('catalogCurrentStatus').textContent=[snapshot.status||'Idle',snapshot.phase,snapshot.execution?.source].filter(Boolean).join(' · ');
    el('catalogCurrentPath').textContent=snapshot.output||'No saved experiment output yet.';
    el('catalogCurrent').disabled=!snapshot.output||busy;
  }
  async function select(source,path){
    busy=true;render();current();el('catalogMessage').textContent='Loading experiment for analysis…';
    try{
      const ok=await window.workshopAnalysis.load(source,path);
      el('catalogMessage').textContent=ok?'Experiment selected for analysis.':'Could not select experiment. See the Analysis message for details.';
      showView('analysis');
    }finally{busy=false;render();current();}
  }
  async function refresh(command={action:'refresh'}){
    if(busy)return;
    busy=true;el('catalogRefresh').disabled=true;el('catalogAdd').disabled=true;render();current();
    el('catalogMessage').textContent=command.action==='add'?'Adding experiment…':'Refreshing catalog…';
    try{
      const data=await request('/api/catalog',command);entries=data.entries;loaded=true;
      el('catalogWarnings').textContent=data.warnings.join(' ');
      el('catalogLocation').textContent=`Catalog: ${data.catalog_path||'not configured'} · Local index: ${data.index_path}`;
      el('catalogMessage').textContent=`Refreshed ${new Date(data.refreshed_at).toLocaleTimeString()}. ${entries.length} entries.`;
      if(command.action==='add')el('catalogAddPath').value='';
    }catch(error){el('catalogMessage').textContent=error.message;}
    finally{busy=false;el('catalogRefresh').disabled=false;el('catalogAdd').disabled=false;render();current();}
  }
  window.addEventListener('workshop-analysis-selected',event=>{
    selectedPath=event.detail.path;el('catalogSelected').textContent=selectedPath||'No experiment selected.';render();
  });
  el('catalogRefresh').onclick=()=>refresh();
  el('catalogAddForm').onsubmit=event=>{event.preventDefault();refresh({action:'add',path:el('catalogAddPath').value});};
  el('catalogSearch').oninput=el('catalogSource').onchange=()=>{page=0;render();};
  el('catalogPrevious').onclick=()=>{page--;render();};el('catalogNext').onclick=()=>{page++;render();};
  el('catalogCurrent').onclick=()=>select('current');
  el('catalogMonitor').onclick=()=>showView('monitor');el('catalogAnalysis').onclick=()=>showView('analysis');
  new MutationObserver(()=>{if(!root.hidden){current();if(!loaded)refresh();}}).observe(root,{attributes:true,attributeFilter:['hidden']});
  // Reuse the workshop state poll; no second acquisition request loop.
  setInterval(()=>{if(!root.hidden&&!document.hidden)current();},1000);
  request('/api/analysis',{action:'status'}).then(data=>{if(!selectedPath&&data.workspace.path){selectedPath=data.workspace.path;el('catalogSelected').textContent=selectedPath;render();}}).catch(()=>{});
  render();current();if(!root.hidden)refresh();
})();

File: braindance/examples/streaming_workshop/catalog_workspace.py
"""Local workshop experiment index alongside the configured recording catalog."""
import json
import threading
from datetime import datetime, timezone
from pathlib import Path

from braindance.config import get_catalog_path, get_data_dir

from .discovery import discover_experiments


class CatalogWorkspace:
    def __init__(self, output_dir):
        self.output_dir = Path(output_dir)
        self.index_path = self.output_dir / 'catalog_experiments.json'
        self.lock = threading.RLock()

    def control(self, command):
        action = command.get('action', 'refresh')
        if action not in ('refresh', 'add'):
            raise ValueError('Unknown catalog action')
        with self.lock:
            saved = json.loads(self.index_path.read_text()) if self.index_path.exists() else []
            if action == 'add':
                if not str(command.get('path') or '').strip():
                    raise ValueError('Enter an experiment directory or recording path')
                path = Path(command['path']).expanduser().resolve()
                if not path.exists():
                    raise ValueError(f'Path does not exist: {path}')
                if not path.is_dir() and path.suffix.lower() not in ('.json', '.h5', '.hdf5', '.npz'):
                    raise ValueError('Choose an experiment directory, experiment JSON, HDF5 or NPZ recording')
                if str(path) not in saved:
                    saved.append(str(path))
                    self.output_dir.mkdir(parents=True, exist_ok=True)
                    temporary = self.index_path.with_suffix('.tmp')
                    temporary.write_text(json.dumps(saved, indent=2))
                    temporary.replace(self.index_path)
            entries, warnings = [], []
            catalog_path = get_catalog_path()
            if catalog_path and Path(catalog_path).is_file():
                try:
                    from braindance.utils.data_manager import load_catalog, Recording
                    catalog_root = Path(catalog_path).parent
                    base_path = catalog_root if (catalog_root / '.complete.json').is_file() else get_data_dir()
                    catalog = load_catalog(base_path=base_path)
                    frame = catalog.df
                    # CSV inference often makes chip IDs numeric; path components are strings.
                    for key in ('proj', 'chip', 'experiment'):
                        if key in frame:
                            frame[key] = frame[key].fillna('').astype(str)
                    metadata = json.loads(frame.to_json(orient='records'))
                    for index, row in enumerate(metadata):
                        rec = Recording(frame.iloc[index], base_path=base_path)
                        try:
                            explicit = str(row.get('full_path') or '')
                            path = Path(explicit).expanduser() if explicit and '://' not in explicit else None
                            if path is not None and not path.is_absolute():
                                path = base_path / path
                            if path is None or not path.exists():
                                path = rec._raw_data_path
                            available = bool(path and path.exists())
                            entries.append(dict(name=str(row.get('experiment') or row.get('exp') or 'Recording'),
                                                project=str(row.get('proj') or ''), chip=str(row.get('chip') or ''),
                                                kind=str(row.get('type') or ''), source='Configured catalog',
                                                path=str(path.resolve()) if available else explicit,
                                                available=available, metadata=row,
                                                reason='' if available else 'Recording is not available locally. Add its local path to analyze it.'))
                        except Exception as exc:
                            warnings.append(f"Could not resolve {row.get('experiment', index)}: {exc}")
                        finally:
                            rec.clear_cache()
                except Exception as exc:
                    warnings.append(f'Could not load configured catalog: {exc}')
            else:
                warnings.append('No configured catalog CSV found. Set its path in Settings; saved workshop experiments are listed below.')

            discovered = {}
            for root, source in ((self.output_dir, 'Workshop run'),
                                 (get_data_dir() / 'tutorials', 'Tutorial data')):
                for experiment in discover_experiments(root):
                    discovered[str(experiment)] = source
            for location in saved:
                path = Path(location)
                if path.name in ('experiment.json', 'experiment_log.json'):
                    path = path.parent
                experiments = discover_experiments(path) if path.is_dir() else []
                for experiment in experiments or [path]:
                    discovered[str(experiment)] = 'Added locally'
            indexed = {entry['path'] for entry in entries if entry['available']}
            for location in sorted(discovered, reverse=True):
                path = Path(location)
                if location in indexed:
                    continue
                entries.append(dict(name=path.name, project='', chip='', kind='Experiment' if path.is_dir() else 'Recording',
                                    path=location, source=discovered[location],
                                    available=path.exists(), metadata={},
                                    reason='' if path.exists() else 'This saved path is no longer available.'))
            return dict(entries=entries, warnings=warnings, catalog_path=str(catalog_path or ''),
                        index_path=str(self.index_path), refreshed_at=datetime.now(timezone.utc).isoformat())
